src/nvim/eval.c

```diff
@@ -9828,34 +9829,15 @@ static void f_getcwd(typval_T *argvars, typval_T *rettv)
     scope_number[i] = argvars[i].vval.v_number;
     // The scope is the current iteration step.
     scope = i;
-
-    if (scope_number[i] < -1) {
-      EMSG(_(e_invarg));
-      return;
-    }
-  }
-
-  // Allocate and initialize the string to return.
-  cwd = xmalloc(MAXPATHL);
-
-  // Get the scope and numbers from the arguments
-  for (int i = 0; i < MAX_CD_SCOPE; i++) {
-    // If there is no argument there are no more scopes after it, break out.
-    if (argvars[i].v_type == VAR_UNKNOWN) {
-      break;
-    }
-    scope_number[i] = argvars[i].vval.v_number;
-    // The scope is the current iteration step.
-    scope = i;
     // It is an error for the scope number to be less than `-1`.
     if (scope_number[i] < -1) {
       EMSG(_(e_invarg));
-      goto end;
+      return;
     }
   }
 
-  // If the deepest scope number is `-1` advance the scope.
-  if (scope_number[scope] < 0) {
+  // If a scope number is `-1` advance the scope by one.
+  while (scope_number[scope] < 0 && scope < MAX_CD_SCOPE) {
     scope++;
```

Comment:
I prefer to avoid nebulous terms like "normalize". What does it even mean to "normalize" a scope? A scope is not a vector where the term would be clearly defined, it's a number out of a finite number set. How about "Skip over `-1` scopes" instead?

You're right about the second one, but after looking at it again it doesn't need to be a loop at all. If all arguments are `-1` the `scope` will already be at its highest value (tab), so it only needs to be incremented to global once. If there are less than maximum arguments (e.g. `haslocaldir(-1)`) then we also only increment the `scope` by one and it will work because the next higher number is `0` (the array of scope numbers has been initialised to all zero).
```
// Numbers of the scope objects (window, tab) we want the working directory
// of. A `-1` means to skip this scope, a `0` means the current object.
int scope_number[] = {
  [kCdScopeWindow] = 0,  // Number of window to look at.
  [kCdScopeTab   ] = 0,  // Number of tab to look at.
};
```

So this will work just fine:
```
if (scope_number[scope] < 0) {
  scope++;
}
```
The problem was that I didn't remember anymore why I had originally written just an `if` check. If it's OK I would rather replace the redundant comment with a comment that explains the *why* instead of the *what*.